### ISSUE TITLE
AWSCognitoAuth refreshToken is nil after a new accessToken is received from the backend

### ISSUE THREAD

**Issue by lucianboboc** [aws/aws-sdk-ios#1035]
### To help us solve your problem better, please answer the following list of questions.

* What service are you using?
AWSCognitoAuth

* In what version of SDK are you facing the problem?
2.6.27

* Is the issue limited to Simulators / Actual Devices?
both

* Can your problem be resolved if you bump to a higher version of SDK?
no

* Is this problem related to specific iOS version?
no

* How are you consuming the SDK? CocoaPods / Carthage / Prebuilt frameworks?
framework and also cocoapods to be able to check the source code and debug.

* Can you give us steps to reproduce with a minimal, complete, and verifiable example? Please include any specific network conditions that might be required to reproduce the problem.
1. Login with gmail.
2. Quit the app and wait 1h till the session expires.
3. Start the app again and call getSession.
4. The refresh token flow is used:
```
NSString *url = [NSString stringWithFormat:@"%@/oauth2/token",self.authConfiguration.webDomain];
                NSString *body = [NSString stringWithFormat:@"grant_type=refresh_token&client_id=%@&refresh_token=%@",self.authConfiguration.appClientId, refreshToken];

```
5. In the `connectionDidFinishLoading:` callback, all keys are received from the server except the `@"refresh_token"` which is missing and the new `AWSCognitoAuthUserSession` object is created with a nil refreshToken value.

If the application is stopped now and started again, the next time the session object is created, the `- (void)getSessionInternal: (nullable UIViewController *) vc completion: (AWSCognitoAuthGetSessionBlock) completion` method is used and the refresh token will be loaded from the keychain:
```
NSString * refreshTokenKey = [self keyChainKey:keyChainNamespace key:AWSCognitoAuthUserRefreshToken];
            NSString * refreshToken = self.keychain[refreshTokenKey];
```

A fix would be possible in the `connectionDidFinishLoading:` to check if the refreshToken is received from the server, otherwise to load it from the keychain.

If you need help with understanding how to implement something in particular then we suggest that you first look into our [developer guide](https://docs.aws.amazon.com/aws-mobile/latest/developerguide/aws-mobile-android-and-iOS.html ). You can also simplify your process of creating an application by using [Mobile Hub](https://console.aws.amazon.com/mobilehub/home#/).

**Comment by desokroshan** [comment]
@lucianboboc, Thank you for reporting the issue. I have filed it as a bug. We will fix it and update this thread once the fix is released.

**Comment by lucianboboc** [comment]
@desokroshan I spent a few hours today debugging the same issue in `AWSCognitoIdentityUser`.
The `-(AWSTask<AWSCognitoIdentityUserSession*> *) getSession` method have the same logic like AWSCognitoAuth, it is provide the session from keychain or it will get a new one if it is expired.
```
NSString * refreshTokenKey = [self keyChainKey:keyChainNamespace key:AWSCognitoIdentityUserRefreshToken];
NSString * refreshToken = self.pool.keychain[refreshTokenKey];

```

In `AWSURLSessionManager` in the `- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)sessionTask didCompleteWithError:(NSError *)error` method the `[[[AWSTask taskWithResult:nil] continueWithSuccessBlock:^id(AWSTask *task)` is called and inside this block, somewhere at the end of the method on the success case where the `delegate.responseObject` is valid the result is set and it has the same problem, the result doesn't have the refresh token.
```
id result = delegate.responseObject;
delegate.taskCompletionSource.result = result;
```

The issue is very similar but it is in a different framework. If this could be fixed by the backend providing the refresh token to in the respose it might be fixed without any changes, otherwise i think both frameworks needs to use the existing refresh token from keychain.

**Comment by behrooziAWS** [comment]
Just so we're clear on the request, for calls to getSession, you want the session to contain the refresh token regardless of whether a new one was issued or not?  And refreshes work, it's just that the session object returned is missing the refresh token correct?

**Comment by lucianboboc** [comment]
@behrooziAWS yes, that’s correct.

**Comment by desokroshan** [comment]
@lucianboboc We have fixed this in 2.6.29. The changes are pushed to GitHub. Please upgrade to the latest and let us know if it fixes the issue.
